source/staking/contracts/Staking.sol

```diff
@@ -0,0 +1,159 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+
+contract Staking is Ownable {
+  using SafeERC20 for IERC20;
+  using SafeMath for uint256;
+  struct Stake {
+    uint256 cliff;
+    uint256 periodLength;
+    uint256 percentPerPeriod;
+    uint256 initialBalance;
+    uint256 currentBalance;
+    uint256 blockNumber;
+  }
+
+  IERC20 public immutable stakingToken;
+  uint256 public cliff;
+  uint256 public periodLength;
+  uint256 public percentPerPeriod;
+  mapping(address => Stake[]) public stakes;
+
+  constructor(
+    IERC20 _stakingToken,
+    uint256 _cliff,
+    uint256 _periodLength,
+    uint256 _percentPerPeriod
+  ) {
+    stakingToken = _stakingToken;
+    cliff = _cliff;
+    periodLength = _periodLength;
+    percentPerPeriod = _percentPerPeriod;
+  }
+
+  function setVestingSchedule(
+    uint256 _cliff,
+    uint256 _periodLength,
+    uint256 _percentPerPeriod
+  ) external onlyOwner {
+    cliff = _cliff;
+    periodLength = _periodLength;
+    percentPerPeriod = _percentPerPeriod;
+  }
+
+  function stake(uint256 amount) external {
+    stakeFor(msg.sender, amount);
+  }
+
+  function stakeFor(address account, uint256 amount) public {
+    stakes[account].push(
+      Stake(cliff, periodLength, percentPerPeriod, amount, amount, block.number)
+    );
+    stakingToken.safeTransferFrom(account, address(this), amount);
+  }
+
+  function addToStake(uint256 index, uint256 amount) external {
+    addToStakeFor(index, msg.sender, amount);
+  }
+
+  function addToStakeFor(
+    uint256 index,
+    address account,
+    uint256 amount
+  ) public {
+    Stake storage stakeData = stakes[msg.sender][index];
```

Comment:
check that stakeData exists

Reply:
added tests for this `27aa2ff6505183ab5855582f76a851d9d6a44e4f`
code reverts when attempting to access an index that doesn't exist.